### PR TITLE
Interest Consensus

### PR DESCRIPTION
**Pull request by hensha256**
### Description

- New - the InterestConsensus contract. This allows nodes to reach a consensus on the accrued interest of a user up to a specific block.
- This pull request therefore also updates the Lenders contract to allow for this new way of updating accrued interest.
- Many tests for Lenders had to be updated/fixed due to the large number of new/deleted functions

Screenshot 2020-04-20 at 10 42 49 AM

### CODE REVIEW

**Comment by salazarguille** [comment]
Please add a description, screenshot for the unit test and code coverage current status.

File: contracts/base/Consensus.sol [hunk 1]
```diff
@@ -39,27 +40,24 @@ contract Consensus is SignerRole {
...
-      uint256 maximum,
-      uint256 minimum,
-      uint256 average
-    ) internal view returns (bool) {
+    function _resultsWithinTolerance(uint256 maximum, uint256 minimum, uint256 average)
+        internal
+        view
+        returns (bool)
+    {
         uint256 toleranceAmount = average.mul(maximumTolerance).div(10000);
```
Comment (salazarguille) [inline contracts/base/Consensus.sol:54]: Please, extract the number to a constant with a description.

File: contracts/base/InterestConsensus.sol [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+contract InterestConsensus is Initializable, Consensus {
+    using SafeMath for uint256;
+
+    LendersInterface public lenderInfo;
+
+    // mapping of (lender, blockNumber) to the aggregated node submissions for their request
+    mapping(address => mapping(uint256 => ZeroCollateralCommon.AggregatedInterest)) nodeSubmissions;
+
+    event InterestSubmitted(
```
Comment (salazarguille) [inline contracts/base/InterestConsensus.sol:38]: Please, consider extracting events/funcntions to an Interface (maybe InterestConsensusInterface / ConsensusInterface).

File: contracts/base/InterestConsensus.sol [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        address signer,
+        address lender,
+        uint256 blockNumber,
+        uint256 interest
+    );
+
```
Comment (salazarguille) [inline contracts/base/InterestConsensus.sol:44]: The same

File: contracts/base/InterestConsensus.sol [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        public
+        Consensus(initRequiredSubmissions, initMaximumTolerance)
+    {}
+
+    function initialize(address lenderInfoAddress) public isNotInitialized() {
+        require(lenderInfoAddress != address(0), "MUST_PROVIDE_LENDER_INFO");
+
+        lenderInfo = LendersInterface(lenderInfoAddress);
+
+        initialize();
```
Comment (salazarguille) [inline contracts/base/InterestConsensus.sol:57]: I think it should be between the "require" and set the "lenderInfo" variable.

File: contracts/base/InterestConsensus.sol [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        nodeSubmissions[lender][blockNumber] = aggregatedData;
+    }
+
+    function _hashData(
+        address lender,
+        uint256 blockNumber,
+        uint256 interest,
+        uint256 signerNonce
+    ) internal pure returns (bytes32) {
+        return keccak256(abi.encode(lender, blockNumber, interest, signerNonce));
```
Comment (salazarguille) [inline contracts/base/InterestConsensus.sol:148]: Have you considered to add address(this) and/or Lenders contract address?

File: contracts/util/ZeroCollateralCommon.sol [hunk 6]
```diff
@@ -2,17 +2,27 @@ pragma solidity 0.5.17;
...
+        uint256 totalNotWithdrawn;
+        uint256 blockLastAccrued;
+    }
+
+    struct AggregatedInterest {
+        uint256 totalSubmissions;
+        uint256 maxValue;
+        uint256 minValue;
+        uint256 sumOfValues;
+        bool finalised;
```
Comment (salazarguille) [inline contracts/util/ZeroCollateralCommon.sol:23]: Would it be 'finalized'?

File: migrations/2_deploy.js [hunk 7]
```diff
@@ -31,19 +28,19 @@ module.exports = async function(deployer, network, accounts) {
   // Creating DeployerApp helper.
   const deployerApp = new DeployerApp(deployer, web3, deployerAccount, network);
 
-  await deployerApp.deploys([ AddressLib, SafeMath, ZDai ], deployOptions);
```
Comment (salazarguille) [inline migrations/2_deploy.js:34]: Why did you remove the libraries?

File: migrations/2_deploy.js [hunk 8]
```diff
@@ -31,19 +28,19 @@ module.exports = async function(deployer, network, accounts) {
...
-  assert(eth_usd, 'Chainlinnk ETH/USD data contract is undefined.');
+  assert(eth_usd, 'Chainlink ETH/USD data contract is undefined.');
   await deployerApp.deploy(EtherUsdAggregator, eth_usd, deployOptions);
 
-  await deployerApp.links(LendingPool, [ AddressLib ]);
   await deployerApp.deploy(LendingPool, deployOptions);
 
-  await deployerApp.links(Lenders, [ AddressLib, SafeMath ]);
-  await deployerApp.deploy(Lenders, ZDai.address, LendingPool.address, deployOptions);
+  await deployerApp.deploy(InterestConsensus, 2, 0, deployOptions);
```
Comment (salazarguille) [inline migrations/2_deploy.js:39]: Please, move values 2 and 0 to env variables.

File: migrations/2_deploy.js [hunk 9]
```diff
@@ -31,19 +28,19 @@ module.exports = async function(deployer, network, accounts) {
...
+  console.log('starting lender info')
```
Comment (salazarguille) [inline migrations/2_deploy.js:40]: Remove console log

File: migrations/2_deploy.js [hunk 10]
```diff
@@ -31,19 +28,19 @@ module.exports = async function(deployer, network, accounts) {
...
 
-  await deployerApp.links(Loans, [ SafeMath ]);
+  await deployerApp.deploy(Lenders, ZDai.address, LendingPool.address, InterestConsensus.address, deployOptions);
+  console.log('finished lender info')
```
Comment (salazarguille) [inline migrations/2_deploy.js:43]: Remove console log

File: migrations/2_deploy.js [hunk 11]
```diff
@@ -31,19 +28,19 @@ module.exports = async function(deployer, network, accounts) {
...
 
-  await deployerApp.links(LendingPool, [ AddressLib ]);
   await deployerApp.deploy(LendingPool, deployOptions);
 
-  await deployerApp.links(Lenders, [ AddressLib, SafeMath ]);
-  await deployerApp.deploy(Lenders, ZDai.address, LendingPool.address, deployOptions);
+  await deployerApp.deploy(InterestConsensus, 2, 0, deployOptions);
+  console.log('starting lender info')
 
-  await deployerApp.links(Loans, [ SafeMath ]);
```
Comment (salazarguille) [inline migrations/2_deploy.js:46]: Why did you remove the link library/contract statement?

File: migrations/2_deploy.js [hunk 12]
```diff
@@ -57,6 +54,11 @@ module.exports = async function(deployer, network, accounts) {
   const zTokenInstance = await ZDai.deployed();
   await zTokenInstance.addMinter(LendingPool.address, { from: deployerAccount });
```
Comment (salazarguille) [inline migrations/2_deploy.js:55]: I think as you removed the LendingPool.zTokenMinted function, this "addMinter" role is not needed now.

File: contracts/base/InterestConsensus.sol (hunk 2 above)
Comment (alicevhenshaw) [reply contracts/base/InterestConsensus.sol:38]: I think it will have to be `InterestConsensusInterface` will do now

File: contracts/util/ZeroCollateralCommon.sol (hunk 6 above)
Comment (alicevhenshaw) [reply contracts/util/ZeroCollateralCommon.sol:23]: Ah yes. `finalised` is the British spelling. I will change it!

File: migrations/2_deploy.js (hunk 7 above)
Comment (alicevhenshaw) [reply migrations/2_deploy.js:34]: You only need to deploy and link a library if it has `external` functions that you use. If all of the functions are internal, then they are copied into your contract upon compilation. You therefore do not need to deploy or link the library.

File: migrations/2_deploy.js (hunk 11 above)
Comment (alicevhenshaw) [reply migrations/2_deploy.js:46]: See above

File: migrations/2_deploy.js (hunk 12 above)
Comment (alicevhenshaw) [reply migrations/2_deploy.js:55]: actually `burn` is marked as `onlyMinter` so we still need the minter role

File: contracts/base/InterestConsensus.sol (hunk 5 above)
Comment (alicevhenshaw) [reply contracts/base/InterestConsensus.sol:140]: Done!
